### PR TITLE
libfetchers: avoid api.github.com ratelimit if no github token is set

### PR DESCRIPTION
**Pull request by Mic92**
If we don't have any github token, we won't be able to fetch private repos, but we are also more likely to run into API limits since we don't have a token. To mitigate this only ever use the github api if we actually have a token.

### CODE REVIEW

**Comment by Mic92** [comment]
Anything else that needs to be done here?

File: src/libfetchers/github.cc [hunk 1]
```diff
@@ -265,14 +265,19 @@ struct GitHubInputScheme : GitArchiveInputScheme
         // FIXME: use regular /archive URLs instead? api.github.com
         // might have stricter rate limits.
         auto host = maybeGetStrAttr(input.attrs, "host").value_or("github.com");
-        auto url = fmt(
+        Headers headers = makeHeadersWithAuthTokens(host);
+        // If we have no auth headers than we default to the public archive
+        // urls so we do not run into rate limits.
+        const auto urlFmt =
```
Comment (pleshevskiy) [inline src/libfetchers/github.cc:271]:
```suggestion
        const auto urlFmt =
            host != "github.com"
            ? "https://%s/api/v3/repos/%s/%s/tarball/%s"
            : headers.empty()
            ? "https://%s/%s/%s/archive/%s.tar.gz"
            : "https://api.%s/repos/%s/%s/tarball/%s"
```

File: src/libfetchers/github.cc [hunk 2]
```diff
@@ -265,14 +265,19 @@ struct GitHubInputScheme : GitArchiveInputScheme
         // FIXME: use regular /archive URLs instead? api.github.com
         // might have stricter rate limits.
         auto host = maybeGetStrAttr(input.attrs, "host").value_or("github.com");
-        auto url = fmt(
+        Headers headers = makeHeadersWithAuthTokens(host);
+        // If we have no auth headers than we default to the public archive
```
Comment (knedlsepp) [inline src/libfetchers/github.cc:269]: than -> then

File: src/libfetchers/github.cc [hunk 3]
```diff
@@ -265,14 +265,19 @@ struct GitHubInputScheme : GitArchiveInputScheme
         // FIXME: use regular /archive URLs instead? api.github.com
```
Comment (knedlsepp) [inline src/libfetchers/github.cc:265]: The FIXME can be dropped now.

**Comment by SuperSandro2000** [comment]
Can this please be merged for the next nix release? Since more and more people switched to flakes we are almost always hitting that rate limit.

**Comment by SuperSandro2000** [comment]
bump

**Comment by Mic92** [comment]
> This seems important to get in, is there anything blocking it?

The lack of Nix maintainers to merge.

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Looks good at a glance, but you'll have to update `tests/github-flakes.nix` as it as a mock of the github API and will need to be extended to support this change.

You can run it with `nix build .\#hydraJobs.tests.githubFlakes`

**Comment by SuperSandro2000** [comment]
https://github.com/Mic92/nix-1/pull/1

**Comment by Mic92** [comment]
The tests have been extended to cover both cases (download via public archive and download via api)

**Review by thufschmitt — APPROVED** [review]
Thanks, looks good :)

And thanks for improving the readability of the github test too :)

(Ideally it'd be nice to serve `github.com` and `api.github.com` separately in it, but that's out of scope here)

**Comment by Kha** [comment]
> To mitigate this only ever use the github api if we actually have a token.

This doesn't seem to be true, the `getRevFromRef` function just above `getDownloadUrl` still uses the API. I just ran into a rate limit error from that. Not sure there's a better way to avoid that than `git ls-remote`.

**Comment by Mic92** [comment]
At least this lets you download archives, which is more often used in CIs than updating flakes.

**Pull request merged**